### PR TITLE
#279: Handling failures during indexing

### PR DESCRIPTION
**Pull request by pgpathan22**
#279

### CODE REVIEW

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt [hunk 1]
```diff
@@ -20,7 +20,7 @@ interface ResourcesIndex {
 
     suspend fun getMeta(id: ResourceId): ResourceMeta
 
-    suspend fun reindex()
+    suspend fun reindex(indexFailedPathCallback: IndexFailedPathCallback)
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt:23]: I think the callback should be replaced with 
`val kindDetectFailedFlow: SharedFlow<Path>`

In `AggregatedResourcesIndex` you just need to merge all the flows from the shards

Also in the presenters in the `onFirstViewAttach` you just need to subscribe to this flow and show toasts

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -80,6 +80,8 @@
     <string name="crash_dialog_description">Sorry, the application crashed. Please send a report to the developers.</string>
     <string name="crash_dialog_title">Crash</string>
     <string name="crash_dialog_comment">You can add a comment here:</string>
+    <string name="toast_could_not_process_link_resource_by_path">Could not process Link resource by path %s</string>
```
Comment (mdrlzy) [inline app/src/main/res/values/strings.xml:83]: I think this applies not only to links
Actually the file name will be enough
And it's better to replace with `Сould not detect kind for $filename`

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -80,6 +80,8 @@
...
+    <string name="toast_could_not_process_resources">Could not process %d resources</string>
```
Comment (mdrlzy) [inline app/src/main/res/values/strings.xml:84]: This resource is not needed, we will show for each file

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/dialog/EditTagsDialogPresenter.kt [hunk 4]
```diff
@@ -50,6 +53,12 @@ class EditTagsDialogPresenter(
     private var wasTextRemovedRecently = false
     private var wasTagRemovedRecently = false
     private var textRemovedRecentlyJob: Job? = null
+    val failedPaths = mutableListOf<Path>()
+    val indexFailedPathCallback = object : IndexFailedPathCallback {
+        override fun indexFailed(path: Path) {
+            failedPaths.add(path)
+        }
+    }
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/dialog/EditTagsDialogPresenter.kt:61]: We do not need to listen fails here, the gallery will handle them

File: app/src/main/java/space/taran/arknavigator/ui/fragments/utils/ToastUtils.kt [hunk 5]
```diff
@@ -22,11 +23,27 @@ fun Context.toastFailedPaths(failedPaths: List<Path>) {
...
+            moreTime = true
+        )
+    } else {
+        toast(
+            R.string.toast_could_not_process_resources,
+            failedPaths.size,
+            moreTime = true
+        )
+    }
+}
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/ui/fragments/utils/ToastUtils.kt:40]: Replace with `toastKindDetectFailedPath` and change the argument to single path

**Review by mdrlzy — CHANGES_REQUESTED** [review]

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/FoldersPresenter.kt [hunk 6]
```diff
@@ -44,7 +46,12 @@ class FoldersPresenter : MvpPresenter<FoldersView>() {
         }
 
     private lateinit var devices: List<Path>
-
+    val failedPaths = mutableListOf<Path>()
+    val indexFailedPathCallback = object : IndexFailedPathCallback {
+        override fun indexFailed(path: Path) {
+            failedPaths.add(path)
+        }
+    }
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/FoldersPresenter.kt:54]: You do not need to collect all the fails, show each.
Same for other screens.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt [hunk 7]
```diff
@@ -19,14 +20,22 @@ data class ResourceMeta(
 
     companion object {
 
-        fun fromPath(path: Path): ResourceMeta? {
+        fun fromPath(
+            path: Path,
+            indexFailedPathCallback: IndexFailedPathCallback
+        ): ResourceMeta? {
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt:26]: You also need to change the result to `Result<ResourceMeta>`

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt (hunk 1 above)
Comment (pgpathan22) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt:23]: I have implemented `SharedFlow` as you suggested and show `toasts` for each fail:
The problem : When multiple failed occurred, the last `toast` will be displayed. others' toasts are quickly disappeared.

Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt:23]: It's okay, we'll replace the regular toast with a custom one

Comment (pgpathan22) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndex.kt:23]: ok

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt [hunk 8]
```diff
@@ -20,31 +20,28 @@ data class ResourceMeta(
...
 
-        fun fromPath(
-            path: Path,
-            indexFailedPathCallback: IndexFailedPathCallback
-        ): ResourceMeta? {
+        fun fromPath(path: Path): Result<ResourceMeta?> {
             val size = Files.size(path)
             if (size < 1) {
-                return null
+                return Result.success(null)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt:26]: `null` doesn't look like success

Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt:26]: Although we were just ignoring it before. But I think we can emit warning that empty file is detected.

**Comment by kirillt** [comment]
Current state:

https://user-images.githubusercontent.com/581023/183119528-24ee318f-37a4-4272-aabc-3a268b59aa3a.mp4

**Comment by kirillt** [comment]
1) Toasts appear only after indexing is over.
2) Toasts appear in groups of 3, not 1 by 1.

https://user-images.githubusercontent.com/581023/183676777-faa0087d-6cac-490e-ace1-db4471f5cf7f.mp4

**Review by mdrlzy — APPROVED** [review]

**Pull request merged**